### PR TITLE
Fix aliasing-based undefined behavior in string functions.

### PR DESCRIPTION
**Pull request by zekoz**
Copies these functions verbatim from musl v1.2.0. #10910  #7279

### CODE REVIEW

**Comment by sbc100** [comment]
Can you mention this in `system/lib/libc/README.md`?

**Comment by zekoz** [comment]
> Can you mention this in `system/lib/libc/README.md`?

Added

File: system/include/libc/features.h [hunk 1]
```diff
@@ -33,4 +33,7 @@
 #define _Noreturn
 #endif
 
+#define weak_alias(old, new) \
+	extern __typeof(old) new __attribute__((__weak__, __alias__(#old)))
```
Comment (sbc100) [inline system/include/libc/features.h:37]: This looks like its already defined in `src/internal/libc.h`..  And in upstream musl it moved to the internal header  `src/include/features.h` (as opposed to this features.h header which is externally visible).

Is this a necessary part of the change?  Do the sources in string not already include `src/internal/libc.h`?

I wonder if its worth bringing over the new `src/include/features.h` wrapper header?

Comment (zekoz) [reply system/include/libc/features.h:37]: Well, you nerd-sniped me. I am seeing how much v1.2.0 I can pull in here.

Comment (sbc100) [reply system/include/libc/features.h:37]: If possible I'd recommend keeping this change focused, unless you want to do a full musl upgrade .. which is quite a big endeavor.  (I think it took weeks or maybe months to completely last time it was done).

Comment (kripken) [reply system/include/libc/features.h:37]: I agree, let's land this now, but @zekoz please submit more PRs with more 1.2 pieces or even the whole thing upgrade!

**Pull request merged**